# Introduction

This work was conducted on the *NeSI* platform, using the programs *wget* and *InSilicoSeq*.

----

### Acquiring genomes

The genomes chosen in the [Introduction notebook](https://github.com/GenomicsAotearoa/metagenomics_summer_school/blob/master/building_resources/0.introduction.md) were simply pulled from the NCBI repository using *wget*.

```
# Pseudomonas aeruginosa
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/152/525/GCA_000152525.1_ASM15252v1/GCA_000152525.1_ASM15252v1_genomic.fna.gz

# Nitrobacter winogradskyi
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/012/725/GCF_000012725.1_ASM1272v1/GCF_000012725.1_ASM1272v1_genomic.fna.gz

# Nitrosomonas europaea
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/009/145/GCF_000009145.1_ASM914v1/GCF_000009145.1_ASM914v1_genomic.fna.gz

# Kuenenia stuttgartiensis
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/900/232/105/GCF_900232105.1_Kuenenia_stuttgartiensis_MBR1/GCF_900232105.1_Kuenenia_stuttgartiensis_MBR1_genomic.fna.gz

# Arcobacter nitrofigilis
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/092/245/GCF_000092245.1_ASM9224v1/GCF_000092245.1_ASM9224v1_genomic.fna.gz

# Desulfovibrio desulfuricans
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/420/465/GCF_000420465.1_ASM42046v1/GCF_000420465.1_ASM42046v1_genomic.fna.gz

# Prochlorococcus marinus
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/015/705/GCA_000015705.1_ASM1570v1/GCA_000015705.1_ASM1570v1_genomic.fna.gz

# Nautilia profundicola
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/021/725/GCF_000021725.1_ASM2172v1/GCF_000021725.1_ASM2172v1_genomic.fna.gz

# Vibrio cholerae
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/745/GCF_000006745.1_ASM674v1/GCF_000006745.1_ASM674v1_genomic.fna.gz

# Staphylococcus aureus
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/027/045/GCA_000027045.1_ASM2704v1/GCA_000027045.1_ASM2704v1_genomic.fna.gz
```

The pPR9 plasmid was downloaded manually, then moved to the same location. Before proceeding, all genomes were unzipped and procesed to have more succinct file and contig names.

The lines in this loop refer to these steps:

1. Unzip the file.
1. Take the first two sets of characters, delimited by the underscore character, and store them to a variable **o**
1. Move the unzipped fasta file to the value stored in **o** suffixed with a standard fasta file extension, *.fna*
1. For each contig in the file, remove any trailing white text in the contig header

```
for i in *.fna.gz;
do
    gunzip ${i}
    o=$(echo ${i} | cut -f1,2 -d '_')
    mv $(basename ${i} .gz) ${o}.fna
    sed -i 's/ .*//g' ${o}.fna
done
```

----

#### Creating a stunted *V. cholerae*

In order to create a *V. cholerae* genome with lateral flagella placement, I took the annotation for the genome and identified the gene coordinates corresponding to the *flhF* and *flhG* genes. These genes were identified by predicting genes from the full genome using *prodigal*, then annotating against the *UniRef100* database that we keep on *NeSI* (see [here](https://github.com/GenomicsAotearoa/environmental_metagenomics/blob/master/metagenomic_annotation/1.gene_annotation.md) for how these commands are run).

I then used a bit of *python* to build a table of each gene in the genome and its coordinates on the parent contig. To begin, I created a dictionary mapping the name of each predicted gene to its start and stop positions.

```
gene_map = {}

for line in open('GCF_000006745.1.prod.faa', 'r'):
    if '>' in line:
        gene, start, stop, *_ = line.split('#')
        gene_name = gene.replace('>', '')
        gene_map[gene_name] = ( int(start), int(stop) )
```

I then created a DataFrame reporting each predicted KO accession to the gene it is assigned. A syntenous block between the two genes of interest was identified and the start/stop coordinates spanning this region were stored.

```
import pandas as pd

kegg_df = pd.read_csv('GCF_000006745.1.prod.kegg', sep='\t')
kegg_df.columns
# Index(['Contig', 'Gene', 'KO'], dtype='object')

kegg_df = kegg_df[ kegg_df.KO.isin(['K02404', 'K04562']) ]
coords = [ v for gene in kegg_df.Gene for v in gene_map[gene] ]

contig_name = kegg_df.Contig.unique()[0]
span_start = min(coords)
span_end = max(coords)
```

This information was then used to identify which contig(s) needed to be split. Because this is a draft genome, it exists in 76 fragments, but due to the organisation of the flagella genes, they both fell on a single contig. I simply looped through the genome to find the contig of interest, the wrote it out as two new pieces.

```
genome_content = open('GCF_000006745.1.fna', 'r').read()

output = open('GCF_000006745.1.sliced.fna', 'w')

for contig_data in genome_content.split('>')[1:]:
    contig, *seq_data = contig_data.split('\n')
    seq_data = ''.join(seq_data)
    if contig.replace('>', '') == contig_name:
        _ = output.write( '>{}_sp1\n{}\n'.format(contig, seq_data[0:span_start]) )
        _ = output.write( '>{}_sp2\n{}\n'.format(contig, seq_data[span_end:]) )
    else:
        _ = output.write( '>{}\n{}\n'.format(contig, seq_data) )

output.close()
```

----

### Building the metagenomes

I decided to use *InSilicoSeq* ([[Escalona et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5224698/)] for creating my mock metagenomes, for the following reasons:

1. It is easy to install using [pip](https://pypi.org/project/pip/)
1. It can handle multiple genomes as input files, and use either complete or draft genomes
1. It contains several inbuilt distributions for creating differential coverage profiles, which is desirable for metagenomic binning

To install the program, a simple *pip* command is all that is needed.

```
pip3 install InSilicoSeq
```

*InSilicoSeq* can take input genomes as either completed (one contig) or draft (multiple contigs). To make it easy to keep track of which genomes were which, I renamed the draft genomes to have the file extension *.fasta*, and left completed genomes with the *.fna* extension. I also needed to concatenate the pPR9 plasmid into the same file as the *S. aureus* genome.

```
mv GCF_000420465.1.fna GCF_000420465.1.fasta
mv GCF_000006745.1.sliced.fna GCF_000006745.1.sliced.fasta

cat GCA_000027045.1.fna GU237136.1.fna > GCA_000027045.1.plasmid.fasta
rm GCA_000027045.1.fna GU237136.1.fna
```

I also found it necessary to isolate genomes *GCF_000012725.1* and *GCF_000009145.1* to a fourth metagenome to ensure their recovery in the binning phase.

```
mv GCF_000012725.1.fna GCF_000012725.1.fa
mv GCF_000009145.1.fna GCF_000009145.1.fa
```

From here, it was simple to pass the genomes into three calls to *InSilicoSeq*.

```
iss generate --draft $(ls *.fasta) \
             --genomes $(ls *.fna) \
             --abundance lognormal -m HiSeq --cpus 8 -o abund_skew1/

iss generate --draft $(ls *.fasta) \
             --genomes $(ls *.fna) \
             --abundance halfnormal -m HiSeq --cpus 8 -o abund_skew2/

iss generate --draft $(ls *.fasta) \
             --genomes $(ls *.fna) \
             --abundance exponential -m HiSeq --cpus 8 -o abund_skew3/

iss generate --genomes $(ls *.fa) \
             --abundance lognormal -m HiSeq --cpus 8 -o abund_skew4/
```

----